internal/grpcsync/pubsub_test.go

```diff
@@ -0,0 +1,211 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package grpcsync
+
+import (
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+type testSubscriber struct {
+	mu      sync.Mutex
+	msgs    []int
+	onMsgCh chan struct{}
+}
+
+func newTestSubscriber(chSize int) *testSubscriber {
+	return &testSubscriber{onMsgCh: make(chan struct{}, chSize)}
+}
+
+func (ts *testSubscriber) OnMessage(msg interface{}) {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+	ts.msgs = append(ts.msgs, msg.(int))
+	select {
+	case ts.onMsgCh <- struct{}{}:
+	default:
+	}
+}
+
+func (ts *testSubscriber) receivedMsgs() []int {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+
+	msgs := make([]int, len(ts.msgs))
+	copy(msgs, ts.msgs)
+
+	return msgs
+}
+
+func (s) TestPubSub_PublishNoMsg(t *testing.T) {
+	pubsub := NewPubSub()
+	defer pubsub.Stop()
+
+	ts := newTestSubscriber(1)
+	pubsub.Subscribe(ts)
+
+	select {
+	case <-ts.onMsgCh:
+		t.Fatalf("Subscriber callback invoked when no message was published")
+	case <-time.After(defaultTestShortTimeout):
+	}
+}
+
+func (s) TestPubSub_PublishMsgs_RegisterSubs_And_Stop(t *testing.T) {
+	pubsub := NewPubSub()
+
+	const numPublished = 10
+
+	ts1 := newTestSubscriber(numPublished)
+	pubsub.Subscribe(ts1)
+	wantMsgs1 := []int{}
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+	// Publish ten messages on the pubsub and ensure that they are received in order by the subscriber.
+	go func() {
+		for i := 0; i < numPublished; i++ {
+			pubsub.Publish(i)
+			wantMsgs1 = append(wantMsgs1, i)
+		}
+		wg.Done()
+	}()
+
+	isTimeout := false
+	go func() {
+		for i := 0; i < numPublished; i++ {
+			select {
+			case <-ts1.onMsgCh:
+			case <-time.After(defaultTestTimeout):
+				isTimeout = true
+			}
+		}
+		wg.Done()
+	}()
+
+	wg.Wait()
+	if isTimeout {
```

Comment:
Why do we fail later in the case of a timeout? I believe its better to t.Fatal whenever the reader goroutine times out and fail fast.

Consider the worst case where onMessage() callback is never invoked. We would have to wait for 10 * 10 ms for the test to fail, where as it could have failed in the first run. 

Let me know if there was a different consideration for why we need the isTimeout boolean to check this. 


--- 

Alternatively, you could use a [t.Errorf()](https://pkg.go.dev/testing#B.Errorf) when a reader loop times out and then do this after the `wg.Wait()`

```
if t.Failed() {
	t.FailNow()
}
```

Reply:
Whatever we choose, there are trade-offs.

`t.Fatal`
If we invoke `t.Fatal()` in the goroutine which is called by the main goroutine, the main isn't cancelled.
So, we can't use this in sub goroutine.

`t.Errof` & `t.Failed()`
It is more like Golang than using boolean flag. But many error messages would be displayed if there is a timeout in loops and it would make logs difficult to read.

Could you please share your opinion about that🙇‍♂️